src/visual.ts

```diff
@@ -260,122 +255,144 @@ export class Visual implements IVisual {
-            minSize,
-            maxSize);
+        const xAxisCardProperties = Helpers.getCategoryAxisProperties(dataView.metadata);
+        const valueAxisProperties = Helpers.getValueAxisProperties(dataView.metadata);
+        const valuesMetadataArray: powerbiVisualsApi.DataViewMetadataColumn[] = Visual.extractValuesMetadata(dataView);
 
-        let xAxisCardProperties: DataViewObject = Helpers.getCategoryAxisProperties(dataView.metadata);
+        const axesLabels = Visual.createAxesLabels(xAxisCardProperties, valueAxisProperties, timeStampColumn.source, valuesMetadataArray);
+        const dots: DataPoint[] = Visual.getDataPointsFromSeries(series);
 
-        let hasDynamicSeries: boolean = !!(timeStampColumn.values && timeStampColumn.source);
-        let hasHighlights: boolean = !!valuesColumn.highlights;
+        behavior.applySelectionStateToData(dots);
 
-        let dataPointLabelSettings = PulseChartDataLabelUtils.getDefaultPulseChartLabelSettings();
-        let gapWidths = Visual.getGapWidths(categoryValues);
-        let maxGapWidth = Math.max.apply(null, gapWidths);
+        return {
+            columns: columns,
+            dots: dots,
+            series: series,
+            isScalar: isScalar,
+            dataLabelsSettings: dataPointLabelSettings,
+            axesLabels: { x: axesLabels.xAxisLabel, y: axesLabels.yAxisLabel },
+            hasDynamicSeries: hasDynamicSeries,
+            categoryMetadata: timeStampColumn.source,
+            categories: categoryValues,
+            settings: settings,
+            grouped: grouped,
+            hasHighlights: !!valuesColumn.highlights,
+            widthOfXAxisLabel: Visual.DefaultXAxisLabelWidth,
+            widthOfTooltipValueLabel: widthOfTooltipValueLabel,
+            heightOfTooltipDescriptionTextLine: heightOfTooltipDescriptionTextLine,
+            runnerCounterHeight: textMeasurementService.measureSvgTextHeight(
+                Visual.GET_RUNNER_COUNTER_TEXT_PROPERTIES("lj", settings.runnerCounter.fontSize.value))
+        };
+    }
 
-        let firstValueMeasureIndex: number = 0, firstGroupIndex: number = 0, secondGroupIndex = 1;
-        let grouped: DataViewValueColumnGroup[] = dataView.categorical.values && dataView.categorical.values.grouped();
-        let y_group0Values = grouped[firstGroupIndex]
+    private static generateSeries(categoryValues: any[], grouped: powerbiVisualsApi.DataViewValueColumnGroup[], settings: PulseChartSettingsModel, columns: DataRoles<powerbiVisualsApi.DataViewValueColumn | powerbiVisualsApi.DataViewCategoryColumn>, timeStampColumn: powerbiVisualsApi.DataViewCategoryColumn, host: IVisualHost, maxCategoryValue: any, minCategoryValue: any, dataPointLabelSettings: ChartDataLabelsSettings, isScalar: boolean, runnerCounterFormatString: string, hasHighlights: boolean, valuesColumn: powerbiVisualsApi.DataViewValueColumn, behavior: Behavior) {
```

Comment:
It's better practice use maximum 3 function parameters. (you can define an object of parameters (and interface for it))